File: 2667/Main.java
import java.util.Collections;
import java.util.Scanner;
import java.util.Queue;
import java.util.LinkedList;
import java.util.List;
import java.util.ArrayList;

public class Main {

	static int mat[][];
	static boolean visited[][];
	static List<Integer> groupList;
	static int N;

	public static void main(String [] args) {

		Scanner sc = new Scanner(System.in);

		N = sc.nextInt();

		mat = new int[N][N];
		visited = new boolean[N][N];
		groupList = new ArrayList<Integer>();

		String str = new String();

		for ( int i = 0 ; i < N ; i++ ) {
			str = sc.next();
			for ( int j = 0; j < N ; j++ ) {
			   mat[i][j] = str.charAt(j) -'0';
			}
		}
	
		for ( int i = 0 ; i < N ; i++ ) {
			for ( int j = 0 ; j < N ; j++ ) {
				if ( mat[i][j] == 0 || visited[i][j] == true) {
					continue;
				}
				else if ( mat[i][j] == 1 && visited[i][j] == false ) {
					BFS(i,j);
				}	
			}
		}

		System.out.println(groupList.size());
		Collections.sort(groupList);

		for ( int i = 0; i < groupList.size() ; i++ ) {
			System.out.println(groupList.get(i));
		}
		

	}
	
	public static void BFS(int x, int y) {
		
		int counter = 1;

		int dx[] = { 1, -1, 0, 0};
		int dy[] = { 0, 0, 1, -1};

		Queue<int[]> queue = new LinkedList<>();

		queue.add( new int[] { x, y } );
		
		visited[x][y] = true;	
			
		while ( !queue.isEmpty() ) {

			int[] pos = queue.poll();
			int newX = pos[0];
			int newY = pos[1];

			for ( int i = 0; i < 4 ; i++ ) {

				int nextX = newX + dx[i];
				int nextY = newY + dy[i];

				if ( nextX >= 0 && nextX < N && nextY >= 0 && nextY < N ) {
					if ( visited[nextX][nextY] == false && mat[nextX][nextY] == 1) {
						counter += 1;
						visited[nextX][nextY] = true;
						queue.add(new int[] {nextX,nextY});
					}					
				}
			}	
		}					
		groupList.add(counter);
	}	
}

File: 2667/main.py
from collections import deque

# up down left right
dx = [0, 0, -1, 1]
dy = [-1, 1, 0, 0]

N = int(input())
mat = [[0 for col in range(N)] for row in range(N)]
visited = [[False for col in range(N)] for row in range(N)]

for col in range(N):
    line = str(input())
    idx = 0
    for row in line:
        mat[col][idx] = int(row)
        idx += 1

def bfs(y,x):
    groups = 1
    queue = deque()
    queue.append((y, x))
    visited[y][x] = True

    while queue:
        y, x = queue.popleft()

        for i in range(4):
            ny = y + dy[i]
            nx = x + dx[i]

            if nx < 0 or ny < 0 or nx >= N or ny >= N:
                continue

            if mat[ny][nx] == 0:
                continue

            if mat[ny][nx] == 1 and visited[ny][nx] is False:
                visited[ny][nx] = True
                groups += 1
                queue.append((ny, nx))

    return groups

groups = []

for col in range(N):
    for row in range(N):
        if mat[col][row] == 1 and visited[col][row] is False:
            num = bfs(col, row)
            groups.append(num)

print(len(groups))
groups.sort()
for houses in groups:
    print(houses)
